firedrake/slate/slac/optimise.py

```diff
@@ -0,0 +1,89 @@
+from gem.node import Memoizer
+from functools import singledispatch
+from contextlib import contextmanager
+import firedrake.slate.slate as sl
+
+
+def optimise(expression):
+    return push_block(expression)
+
+
+def push_block(expression):
+    mapper = Memoizer(_push_block)
+    mapper.block = None
+    ret = mapper(expression)
+    return ret
+
+
+@singledispatch
+def _push_block(expr, self):
+    raise AssertionError("Cannot handle terminal type: %s" % type(expr))
+
+
+@_push_block.register(sl.Transpose)
+def _push_block_transpose(expr, self):
+    if self.block:
+        with block(self, *expr.children, self.block._indices[::-1]):
+            ret = sl.Transpose(self(*expr.children))
+    else:
+        ret = expr
+    return ret
+
+
+@_push_block.register(sl.Add)
+def _push_block_add(expr, self):
+    if self.block:
+        ops = []
+        for op in expr.children:
+            with block(self, op, self.block._indices):
+                ops += [self(op)]
```

Comment:
Is there a way to do line 36-40 in one line? I didn't how I can use the context manager and map at the same time.

Reply:
I think this is a slightly weird model. I think this is a use-case for `MemoizerArg` rather than `Memoizer` because you want to use the call stack to maintain the current indices to inline.

For inspiration, I would look at the implementation of `remove_componenttensors` in `tsfc/gem/optimise.py`

Reply:
OK will do, thanks.